### PR TITLE
SHARE-124 - DB migration and implementation of UserAPI

### PR DESCRIPTION
**Pull request by bonbuo**
added DB migration for ranking_score and implemented getUsers function

---

### Your checklist for this pull request

Please review the [contributing guidelines](https://github.com/Catrobat/Catroweb-Symfony/blob/develop/.github/contributing.md) and [wiki pages](https://github.com/Catrobat/catroweb-Symfony/wiki/) of this repository.

- [ ] Include the name and id of the Jira ticket in the PR’s title eg.: `SHARE-666 The devils ticket`
- [ ] Choose the proper base branch (_develop_)
- [ ] Confirm that the changes follow the project’s coding guidelines
- [ ] Verify that the changes generate no warnings and errors
- [ ] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [ ] Verify that all tests are passing (CI), if not please state the test cases in the [section](#Tests) below
- [ ] Perform a self-review of the changes
- [ ] Stick to the project’s git workflow (rebase and squash your commits)
- [ ] Verify that your changes do not have any conflicts with the base branch
- [ ] Put your ticket into the `Code Review` section in [Jira](https://jira.catrob.at/)
- [ ] Post a message in the _#catroweb_ [Slack channel](https://catrobat.slack.com) and ask for a code reviewer
- [ ] Check that your pull request has been successfully deployed to https://web-test-1.catrob.at/

### Additional Description

`TODO: Add additional information that is not in your commit-message here`

### Tests - additional information

`TODO: add additional information about testruns here`

### CODE REVIEW

**Comment by dmetzner** [comment]
@bonbuo please rebase - I have added your version + the stub to develop already; Otherwise you would merge conflicts in the composer files all the time. That's the reason why it's really helpful to create multiple smaller pull requests. Something that could also be easily done for the DB migration :)

File: migrations/2023/Version20231220122134.php [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+    public function up(Schema $schema): void
+    {
+        // this up() migration is auto-generated, please modify it to your needs
+        $this->addSql('ALTER TABLE fos_user ADD ranking_score INT DEFAULT NULL');
+        $this->addSql('CREATE UNIQUE INDEX UNIQ_957A64793F5D60E0 ON fos_user (ranking_score)');
+        $this->addSql('ALTER TABLE studio_activity CHANGE type type ENUM(\'comment\', \'project\', \'user\')');
+        $this->addSql('ALTER TABLE studio_user CHANGE role role ENUM(\'admin\', \'member\'), CHANGE status status ENUM(\'active\', \'banned\', \'pending_request\')');
+        $this->addSql('ALTER TABLE user_like_similarity_relation CHANGE similarity similarity NUMERIC(4, 3) DEFAULT \'0\' NOT NULL');
+        $this->addSql('ALTER TABLE user_remix_similarity_relation CHANGE similarity similarity NUMERIC(4, 3) DEFAULT \'0\' NOT NULL');
```
Comment (dmetzner) [inline migrations/2023/Version20231220122134.php]: those lines are noise and can be removed

File: migrations/2023/Version20231220122134.php [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    public function down(Schema $schema): void
+    {
+        // this down() migration is auto-generated, please modify it to your needs
+        $this->addSql('ALTER TABLE user_like_similarity_relation CHANGE similarity similarity NUMERIC(4, 3) DEFAULT \'0.000\' NOT NULL');
+        $this->addSql('ALTER TABLE user_remix_similarity_relation CHANGE similarity similarity NUMERIC(4, 3) DEFAULT \'0.000\' NOT NULL');
```
Comment (dmetzner) [inline migrations/2023/Version20231220122134.php]: those lines are noise and can be removed

File: migrations/2023/Version20231220122134.php [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        $this->addSql('DROP INDEX UNIQ_957A64793F5D60E0 ON fos_user');
+        $this->addSql('ALTER TABLE fos_user DROP ranking_score');
+        $this->addSql('ALTER TABLE studio_user CHANGE role role VARCHAR(255) DEFAULT NULL, CHANGE status status VARCHAR(255) DEFAULT NULL');
+        $this->addSql('ALTER TABLE studio_activity CHANGE type type VARCHAR(255) DEFAULT NULL');
```
Comment (dmetzner) [inline migrations/2023/Version20231220122134.php]: those lines are noise and can be removed

File: src/Api/Services/User/UserResponseManager.php [hunk 4]
```diff
@@ -28,7 +28,7 @@ public function createBasicUserDataResponse(User $user, string $attributes = nul
     if (empty($attributes)) {
       $attributes_list = ['id', 'username'];
     } elseif ('ALL' === $attributes) {
-      $attributes_list = ['id', 'username', 'picture', 'about', 'currentlyWorkingOn', 'projects', 'followers', 'following'];
+      $attributes_list = ['id', 'username', 'picture', 'about', 'currentlyWorkingOn', 'projects', 'followers', 'following', 'ranking_score'];
```
Comment (dmetzner) [inline src/Api/Services/User/UserResponseManager.php:31]: we should try to keep a consistent naming convention;

currentlyWorkingOn vs ranking_score

Comment (dmetzner) [reply src/Api/Services/User/UserResponseManager.php:31]: Maybe you could change `currentlyWorkingOn`; pretty sure no one is using it already, but please double check

File: src/System/Commands/DBUpdater/UpdateUserRankingCommand.php [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+     * @param OutputInterface $output
+     * @return int
+     */
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $output->writeln('Recomputing ELO ranking for all users');
+
+        $users = $this->userRepository->findAll();
+
+        foreach ($users as $user) {
```
Comment (dmetzner) [inline src/System/Commands/DBUpdater/UpdateUserRankingCommand.php:47]: consider doing the computation in batches, we have a lot of users.

File: src/Api/Services/User/UserResponseManager.php (hunk 4 above)
Comment (bonbuo) [reply src/Api/Services/User/UserResponseManager.php:31]: it seems there are some translations and already some usages for currentlyWorkingOn

Comment (dmetzner) [reply src/Api/Services/User/UserResponseManager.php:31]: yes, but only on our side, it is not a feature used by the apps yet :)
Can you please fix this in a new merge request?

**Review by dmetzner — APPROVED** [review]

**Pull request merged**